File: reference.vb.net/Db4objects.Db4odoc/Replication/ReplicationExample.vb
Imports System
Imports Db4objects.Db4o
Imports Db4objects.Db4o.Config
Imports Db4objects.Db4o.Query
Imports Db4objects.Drs


Namespace Db4objects.Db4odoc.Replicating
    Public Class ReplicationExample
        Public Shared ReadOnly DtFileName As String = "reference.db4o"
        Public Shared ReadOnly HhFileName As String = "handheld.db4o"

        Public Shared Sub ConfigureReplication()
            Db4oFactory.Configure().GenerateUUIDs(ConfigScope.GLOBALLY)
            Db4oFactory.Configure().GenerateVersionNumbers(ConfigScope.GLOBALLY)
        End Sub
        ' end configureReplication

        Public Shared Sub ConfigureReplicationPilot()
            Db4oFactory.Configure().ObjectClass(GetType(Pilot)).GenerateUUIDs(True)
            Db4oFactory.Configure().ObjectClass(GetType(Pilot)).GenerateVersionNumbers(True)
        End Sub
        ' end configureReplicationPilot

        Public Shared Sub ConfigureForExisting()
            Db4oFactory.Configure().ObjectClass(GetType(Pilot)).EnableReplication(True)
            Defragment.Defragment.Defrag(DtFileName)
        End Sub
        ' end configureForExisting

        Public Shared Sub Replicate()
            Dim desktop As IObjectContainer = Db4oFactory.OpenFile(DtFileName)
            Dim handheld As IObjectContainer = Db4oFactory.OpenFile(HhFileName)
            Dim replic As IReplicationSession = Replication.Begin(handheld, desktop) '
            ' There is no need to replicate all the objects each time. 
            ' ObjectsChangedSinceLastReplication methods gives us 
            ' a list of modified objects
            Dim changed As IObjectSet = replic.ProviderA().ObjectsChangedSinceLastReplication()
            'Iterate through the changed objects, replicate them
            While changed.HasNext()
                replic.Replicate(changed.Next())
            End While
            replic.Commit()
        End Sub
        ' end replicate	

        Public Shared Sub ReplicatePilots()
            Dim desktop As IObjectContainer = Db4oFactory.OpenFile(DtFileName)
            Dim handheld As IObjectContainer = Db4oFactory.OpenFile(HhFileName)
            Dim replic As IReplicationSession = Replication.Begin(handheld, desktop) '
            ' There is no need to replicate all the objects each time. 
            ' ObjectsChangedSinceLastReplication methods gives us 
            ' a list of modified objects
            Dim changed As IObjectSet = replic.ProviderB().ObjectsChangedSinceLastReplication()
            ' Iterate through the changed objects,
            ' check if the name starts with "S" and replicate only those items
            While changed.HasNext()
                Dim p As Object = changed.Next
                If (p Is GetType(Pilot)) Then
                    If (CType(p, Pilot)).Name.StartsWith("S") Then
                        replic.Replicate(p)
                    End If
                End If
            End While
            replic.Commit()
        End Sub
        ' end ReplicatePilots

        Public Shared Sub ReplicateBiDirectional()
            Dim desktop As IObjectContainer = Db4oFactory.OpenFile(DtFileName)
            Dim handheld As IObjectContainer = Db4oFactory.OpenFile(HhFileName)
            Dim replic As IReplicationSession = Replication.Begin(handheld, desktop) '
            Dim changed As IObjectSet = replic.ProviderA().ObjectsChangedSinceLastReplication()
            'Iterate changed objects, replicate them
            While changed.HasNext()
                replic.Replicate(changed.Next())
            End While

            changed = replic.ProviderB().ObjectsChangedSinceLastReplication()
            ' Add one more loop for bi-directional replication
            While changed.HasNext()
                replic.Replicate(changed.Next())
            End While
            replic.Commit()
        End Sub
        ' end ReplicateBiDirectional
    End Class
End Namespace

File: reference.net/db4o-reference-chapters/src/Db4objects.Db4odoc/Replication/ReplicationExample.cs
using System;

using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Query;

using Db4objects.Drs;

namespace Db4objects.Db4odoc.Replication
{
    public class ReplicationExample
    {
        //private const string DtFileName = "reference.db4o";
        //private const string HhFileName = "handheld.db4o";

        //public static void ConfigureReplication()
        //{
        //    Db4oFactory.Configure().GenerateUUIDs(ConfigScope.GLOBALLY);
        //    Db4oFactory.Configure().GenerateVersionNumbers(ConfigScope.GLOBALLY);
        //}
        //// end configureReplication

        //public static void ConfigureReplicationPilot()
        //{
        //    Db4oFactory.Configure().ObjectClass(typeof(Pilot)).GenerateUUIDs(true);
        //    Db4oFactory.Configure().ObjectClass(typeof(Pilot)).GenerateVersionNumbers(true);
        //}
        //// end configureReplicationPilot

        //public static void ConfigureForExisting()
        //{
        //    Db4oFactory.Configure().ObjectClass(typeof(Pilot)).EnableReplication(true);
        //    Db4objects.Db4o.Defragment.Defragment.Defrag(DtFileName);
        //}
        //// end configureForExisting

        //public static void Replicate()
        //{
        //    IObjectContainer desktop = Db4oFactory.OpenFile(DtFileName);
        //    IObjectContainer handheld = Db4oFactory.OpenFile(HhFileName);
        //    IReplicationSession replication = Db4objects.Drs.Replication.Begin(handheld, desktop);
            
        //    /*
        //     * There is no need to replicate all the objects each time. 
        //     * ObjectsChangedSinceLastReplication methods gives us 
        //     * a list of modified objects
        //     */
        //    IObjectSet changed = replication.ProviderA().ObjectsChangedSinceLastReplication();
        //    //Iterate through changed objects, replicate them
        //    while (changed.HasNext())
        //    {
        //        replication.Replicate(changed .Next ());
        //    }
        //    replication.Commit();
        //}
        //// end replicate	

        //public static void ReplicatePilots()
        //{
        //    IObjectContainer desktop = Db4oFactory.OpenFile(DtFileName);
        //    IObjectContainer handheld = Db4oFactory.OpenFile(HhFileName);
        //    IReplicationSession replication = Db4objects.Drs.Replication.Begin(handheld, desktop);
        //    IObjectSet changed = replication.ProviderB().ObjectsChangedSinceLastReplication();
        //    //Iterate changed objects, replicate them
        //    while (changed.HasNext())
        //    {
        //        object p = changed .Next();
        //        if ( p is Pilot)
        //        {
        //            if (((Pilot)p).Name.StartsWith("S"))
        //            {
        //                replication.Replicate(p);
        //            }
        //        }
        //    }
        //    replication.Commit();
        //}
        //// end ReplicatePilots

        //public static void ReplicateBiDirectional()
        //{
        //    IObjectContainer desktop = Db4oFactory.OpenFile(DtFileName);
        //    IObjectContainer handheld = Db4oFactory.OpenFile(HhFileName);
        //    IReplicationSession replication = Db4objects.Drs.Replication.Begin(handheld, desktop);
        //    IObjectSet changed = replication.ProviderA().ObjectsChangedSinceLastReplication();
        //    while (changed.HasNext())
        //    {
        //        replication.Replicate(changed.Next());
        //    }

        //    // Add one more loop for bi-directional replication
        //    changed = replication.ProviderB().ObjectsChangedSinceLastReplication();
        //    while (changed.HasNext())
        //    {
        //        replication.Replicate(changed.Next());
        //    }
        //    replication.Commit();
        //}
        // end ReplicateBiDirectional
    }

}
